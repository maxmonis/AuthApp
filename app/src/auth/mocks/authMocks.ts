export const mockEmail = "user@example.com"
export const mockPassword = "Passw0rd"
export const mockToken = "mock-token"
export const mockRefreshedToken = "mock-refreshed-token"
