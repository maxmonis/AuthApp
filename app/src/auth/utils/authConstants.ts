export const accountRoutes = [
  "forgot-password",
  "login",
  "register",
  "update-password",
] as const
