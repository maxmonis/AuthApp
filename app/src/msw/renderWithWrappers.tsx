import {render} from "@testing-library/react"
import {QueryClient, QueryClientProvider} from "react-query"
import {BrowserRouter} from "react-router-dom"

const queryClient = new QueryClient({defaultOptions: {queries: {retry: false}}})

export function renderWithWrappers(element: React.ReactNode) {
  return render(
    <QueryClientProvider client={queryClient}>
      <BrowserRouter>{element}</BrowserRouter>
    </QueryClientProvider>,
  )
}
