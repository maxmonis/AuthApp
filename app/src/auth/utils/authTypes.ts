import {accountRoutes} from "./authConstants"

export type AccountRoute = (typeof accountRoutes)[number]

export type TokenRes = {token: string}

export type UserRes = {email: string}
