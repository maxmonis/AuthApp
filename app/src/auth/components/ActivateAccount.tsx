import axios from "axios"
import React from "react"
import {Link, useNavigate, useParams} from "react-router-dom"
import {AccountLayout} from "src/auth/components/AccountLayout"
import {useMounted} from "src/hooks/useMounted"
import {useAuth} from "../hooks/useAuth"
import {useUpdateAuth} from "../hooks/useUpdateAuth"
import {TokenRes} from "../utils/authTypes"

export function ActivateAccount() {
  const navigate = useNavigate()
  const {token} = useParams<{token: string}>()
  const {data: user} = useAuth()
  const updateAuth = useUpdateAuth()
  const [error, setError] = React.useState(false)
  useMounted(async () => {
    try {
      const res = await axios.post<TokenRes>("/api/user/activate", {token})
      await updateAuth(res.data.token)
      navigate("/", {replace: true})
    } catch (error) {
      console.error(error)
      setError(true)
    }
  })
  if (!error) return <></>
  return (
    <AccountLayout title="Invalid Link">
      <p>We were unable to activate your account using the provided link.</p>
      <p>
        If you've already completed setup,{" "}
        <Link
          className="whitespace-nowrap text-blue-600 dark:text-blue-400"
          replace
          to={user ? "/" : "/login"}
        >
          visit your account
        </Link>
        .
      </p>
      <p>
        If your activation link has expired,{" "}
        <Link
          className="whitespace-nowrap text-blue-600 dark:text-blue-400"
          onClick={() => updateAuth(null)}
          replace
          to="/register"
        >
          sign up again
        </Link>
        .
      </p>
    </AccountLayout>
  )
}
