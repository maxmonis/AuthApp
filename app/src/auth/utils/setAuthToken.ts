import axios from "axios"
import {localToken} from "src/utils/storage"
import {TokenRes} from "./authTypes"

export function setAuthToken({refresh = false, token = localToken.get()}) {
  if (token) {
    axios.defaults.headers.common["X-Auth-Token"] = token
    localToken.set(token)
    if (refresh)
      axios
        .get<TokenRes>("/api/user/refresh")
        .then(res => setAuthToken({token: res.data.token}))
        .catch(() => setAuthToken({token: null}))
  } else {
    delete axios.defaults.headers.common["X-Auth-Token"]
    localToken.remove()
  }
}
