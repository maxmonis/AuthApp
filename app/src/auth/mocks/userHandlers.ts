import {rest} from "msw"
import {mockDelay} from "src/msw/mockDelay"
import {TokenRes, UserRes} from "../utils/authTypes"
import {
  mockEmail,
  mockPassword,
  mockRefreshedToken,
  mockToken,
} from "./authMocks"

export const userHandlers = [
  /* -------------------- Load User -------------------- */
  rest.get("api/user", (req, res, ctx) => {
    if (!req.headers.has("X-Auth-Token"))
      return res(ctx.status(401), ctx.json("No token"), ctx.delay(mockDelay()))
    if (
      ![mockToken, mockRefreshedToken].includes(
        req.headers.get("X-Auth-Token") ?? "",
      )
    )
      return res(
        ctx.status(401),
        ctx.json("Invalid token"),
        ctx.delay(mockDelay()),
      )
    return res(
      ctx.json({email: mockEmail} satisfies UserRes),
      ctx.delay(mockDelay()),
    )
  }),

  /* -------------------- Refresh Token -------------------- */
  rest.get("api/user/refresh", (req, res, ctx) => {
    if (!req.headers.has("X-Auth-Token"))
      return res(ctx.status(401), ctx.json("No token"), ctx.delay(mockDelay()))
    if (
      ![mockToken, mockRefreshedToken].includes(
        req.headers.get("X-Auth-Token") ?? "",
      )
    )
      return res(
        ctx.status(401),
        ctx.json("Invalid token"),
        ctx.delay(mockDelay()),
      )
    return res(
      ctx.json({token: mockRefreshedToken} satisfies TokenRes),
      ctx.delay(mockDelay()),
    )
  }),

  /* -------------------- Log In -------------------- */
  rest.post("api/user/login", async (req, res, ctx) => {
    const {email, password} = await req.json()
    if (email !== mockEmail)
      return res(
        ctx.status(404),
        ctx.json("User not found"),
        ctx.delay(mockDelay()),
      )
    if (password !== mockPassword)
      return res(
        ctx.status(401),
        ctx.json("Incorrect password"),
        ctx.delay(mockDelay()),
      )
    return res(
      ctx.json({token: mockToken} satisfies TokenRes),
      ctx.delay(mockDelay()),
    )
  }),

  /* -------------------- Create Account -------------------- */
  rest.post("api/user/create", async (req, res, ctx) => {
    const {email} = await req.json()
    if (email === mockEmail)
      return res(
        ctx.status(400),
        ctx.json("User already exists"),
        ctx.delay(mockDelay()),
      )
    return res(ctx.json("Link email sent"), ctx.delay(mockDelay()))
  }),
  rest.post("api/user/activate", async (req, res, ctx) => {
    const {token} = await req.json()
    if (token !== mockToken)
      return res(
        ctx.status(401),
        ctx.json("Invalid token"),
        ctx.delay(mockDelay()),
      )
    return res(
      ctx.json({token: mockToken} satisfies TokenRes),
      ctx.delay(mockDelay()),
    )
  }),

  /* -------------------- Update Password -------------------- */
  rest.post("api/user/update-password", async (req, res, ctx) => {
    const {email} = await req.json()
    if (email !== mockEmail)
      return res(
        ctx.status(404),
        ctx.json("User not found"),
        ctx.delay(mockDelay()),
      )
    return res(ctx.json("Password update email sent"), ctx.delay(mockDelay()))
  }),
  rest.patch("api/user/update-password", async (req, res, ctx) => {
    const {token} = await req.json()
    if (token !== mockToken)
      return res(
        ctx.status(401),
        ctx.json("Invalid token"),
        ctx.delay(mockDelay()),
      )
    return res(
      ctx.json({token: mockToken} satisfies TokenRes),
      ctx.delay(mockDelay()),
    )
  }),
]
