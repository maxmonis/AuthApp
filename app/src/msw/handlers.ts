import {userHandlers} from "src/auth/mocks/userHandlers"

export const handlers = [...userHandlers]
