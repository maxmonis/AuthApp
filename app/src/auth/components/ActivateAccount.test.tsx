import {screen, waitFor} from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import {renderWithWrappers} from "src/msw/renderWithWrappers"
import {localToken} from "src/utils/storage"
import {mockToken} from "../mocks/authMocks"
import {setAuthToken} from "../utils/setAuthToken"
import {ActivateAccount} from "./ActivateAccount"

const mockNavigate = jest.fn()
let mockParams = {token: "invalid-token"}
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockNavigate,
  useParams: () => mockParams,
}))

describe("ActivateAccount", () => {
  test("shows message if token invalid", async () => {
    const {asFragment} = renderWithWrappers(<ActivateAccount />)
    await screen.findByText(
      "We were unable to activate your account using the provided link.",
    )
    expect(asFragment()).toMatchSnapshot()
  })
  test("message includes link to login if no user is signed in", async () => {
    renderWithWrappers(<ActivateAccount />)
    await screen.findByText(
      "We were unable to activate your account using the provided link.",
    )
    expect(
      screen.getByRole("link", {name: "visit your account"}),
    ).toHaveAttribute("href", "/login")
  })
  test("message includes link to home if user is signed in", async () => {
    setAuthToken({token: mockToken})
    renderWithWrappers(<ActivateAccount />)
    await screen.findByText(
      "We were unable to activate your account using the provided link.",
    )
    expect(
      screen.getByRole("link", {name: "visit your account"}),
    ).toHaveAttribute("href", "/")
  })
  test("message includes link to register", async () => {
    renderWithWrappers(<ActivateAccount />)
    await screen.findByText(
      "We were unable to activate your account using the provided link.",
    )
    expect(screen.getByRole("link", {name: "sign up again"})).toHaveAttribute(
      "href",
      "/register",
    )
  })
  test("clicking register link logs user out", async () => {
    setAuthToken({token: mockToken})
    renderWithWrappers(<ActivateAccount />)
    await screen.findByText(
      "We were unable to activate your account using the provided link.",
    )
    userEvent.click(screen.getByRole("link", {name: "sign up again"}))
    expect(localToken.get()).toBeNull()
  })
  test("navigates to home page on success", async () => {
    mockParams = {token: mockToken}
    renderWithWrappers(<ActivateAccount />)
    await waitFor(() => {
      expect(mockNavigate).toHaveBeenCalledWith("/", {replace: true})
    })
  })
})
