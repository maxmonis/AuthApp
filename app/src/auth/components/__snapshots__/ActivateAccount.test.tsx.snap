// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivateAccount shows message if token invalid 1`] = `
<DocumentFragment>
  <div
    class="flex justify-center px-6"
  >
    <div
      class="mb-40 mt-10 flex max-h-min w-full max-w-xs flex-col gap-3 rounded-md border border-slate-700 bg-white p-6 dark:bg-black"
    >
      <div
        class="mx-auto flex items-center gap-2"
      >
        <img
          alt="AuthApp logo"
          class="h-6 w-6"
          src="/logo192.png"
        />
        <h1
          class="text-2xl"
        >
          AuthApp
        </h1>
      </div>
      <h2
        class="text-center text-xl font-bold"
      >
        Invalid Link
      </h2>
      <p>
        We were unable to activate your account using the provided link.
      </p>
      <p>
        If you've already completed setup, 
        <a
          class="whitespace-nowrap text-blue-600 dark:text-blue-400"
          href="/login"
        >
          visit your account
        </a>
        .
      </p>
      <p>
        If your activation link has expired, 
        <a
          class="whitespace-nowrap text-blue-600 dark:text-blue-400"
          href="/register"
        >
          sign up again
        </a>
        .
      </p>
    </div>
  </div>
</DocumentFragment>
`;
